storey/redis_driver.py

```diff
@@ -0,0 +1,482 @@
+        self._aggregation_time_attribute_prefix = aggregation_time_attribute_prefix
+        self._aggregation_prefixes = (self._aggregation_attribute_prefix,
+                                      self._aggregation_time_attribute_prefix)
+
+    def asyncify(self, fn):
+        """Run a synchronous function asynchronously."""
+        async def inner_fn(*args, **kwargs):
+            loop = asyncio.get_event_loop()
+            partial_hset = partial(fn, *args, **kwargs)
+            return await loop.run_in_executor(RedisDriver._thread_pool, partial_hset)
+        return inner_fn
+
+    @property
+    def redis(self):
+        if self._redis:
+            return self._redis
+        if self._type is RedisType.STANDALONE:
+            return redis.Redis.from_url(self._redis_url, decode_responses=True)
+        self._redis = rediscluster.RedisCluster.from_url(self._redis_url, decode_response=True)
+        return self._redis
+
+    @staticmethod
+    def make_key(key_prefix, *parts):
+        return f"{key_prefix}{''.join([str(p) for p in parts])}"
+
+    def _make_key(self, *parts):
+        return RedisDriver.make_key(self._key_prefix, *parts)
+
+    @staticmethod
+    def _static_data_key(redis_key_prefix):
+        """
+        The Redis key for the Hash containing static data (AKA, "additional data").
+
+        The `redis_key_prefix` string should be the Redis key prefix used to store
+        data for a Storey container, table path, and key combination.
+        """
+        return f"{redis_key_prefix}:static"
+
+    def _aggregation_time_key(self, redis_key_prefix, feature_name):
+        """
+        The Redis key containing the associated timestamp for an aggregation.
+
+        The `redis_key_prefix` string should be the Redis key prefix used to store
+        data for a Storey container, table path, and key combination.
+        """
+        return f"{redis_key_prefix}:{self._aggregation_time_attribute_prefix}{feature_name}"
+
+    @staticmethod
+    def _list_key(redis_key_prefix, list_attribute_name):
+        """
+        The key containing a list of aggregation data.
+
+        The `redis_key_prefix` string should be the Redis key prefix used to store
+        data for a Storey container, table path, and key combination.
+        """
+        return f"{redis_key_prefix}:{list_attribute_name}"
+
+    @classmethod
+    def _convert_python_obj_to_redis_value(cls, value):
+        if isinstance(value, datetime):
+            if pd.isnull(value):
+                return None
+            return f'{cls.DATETIME_FIELD_PREFIX}{value.timestamp()}'
+        elif isinstance(value, timedelta):
+            return f'{cls.TIMEDELTA_FIELD_PREFIX}{value.total_seconds()}'
+        else:
+            if isinstance(value, float):
+                if value == math.inf or value == -math.inf:
+                    # Use the shorter infinity form (inf, -inf) to save space.
+                    value = str(value)
```

Comment:
```
>>> json.dumps(math.inf)
'Infinity'
>>> json.dumps(str(math.inf))
'"inf"'
```
The comment implies that `value = str(value)` is an optimization, but it the code doesn't work without it. So need to remove or improve the comment.

What happens if the value is NaN?

Reply:
yep, found it in test_write_read_first_last() checks - will be addressed as part of it.